### ISSUE TITLE
Linux: ucontext_t definition is only correct for glibc 2.28 and above

### ISSUE THREAD

**Issue by acfoltzer** [rust-lang/libc#1410]
This commit introduced a shadow stack `__ssp` field to the definition of `ucontext_t`: https://github.com/rust-lang/libc/commit/4497a7867d56ef61194e4cbb63ef4c4fe146e4c6

However, that field is only present in glibc 2.28 and above, so this struct doesn't work correctly on distros with earlier versions, including Ubuntu 18.04 where I encountered memory corruption after updating from 0.2.47 to 0.2.58.

**Comment by gnzlbg** [comment]
How is code compiled for glibc < 2.28 ABI compatible with >= 2.28 ?

**Comment by acfoltzer** [comment]
Sorry, I must not be understanding your question. I'm on a < 2.28 distro, compiling and running against < 2.28; I currently don't use any >= 2.28 distros.

**Comment by gnzlbg** [comment]
If you compile a binary against glibc < 2.28 headers, ucontext_t has a smaller size than if one uses glibc >= 2.28. AFAIK, you can compile against glibc < 2.28 and then use that binary in a system where glibc >= 2.28 is dynamically linked. I was wondering how does this work here?

E.g., `getcontext` on glibc >= 2.28 might try to write to the extra field of `ucontext_t`, but if your library used older headers, it wouldn't have the field, so that would be a write out-of-bounds. 

Maybe there are new different symbols for newer `getcontext`, so that if you compile your binary against an older libc, and dynamically link against a new version, the older symbols get picked, and these do not write to the extra field ? If so, then we could fix this by using `link_name` to make these APIs link to their older symbols.

**Comment by gnzlbg** [comment]
Relevant commit: https://github.com/bminor/glibc/commit/25123a1c5c96429d70e75b85a9749b405909d7f2

So apparently that's allowed, see: https://github.com/bminor/glibc/commit/25123a1c5c96429d70e75b85a9749b405909d7f2

**Comment by gnzlbg** [comment]
Do you have a minimal working example showing the issue ?

**Comment by gnzlbg** [comment]
(This issue might be related: https://github.com/nix-rust/nix/issues/1092)

**Comment by acfoltzer** [comment]
So, this is coming up in a case where a Rust program is writing a `libc::ucontext_t` to a pointer to a `ucontext_t` provided by C. Since the size of the struct is smaller on the C side, when we write to it using the Rust type we write off the end into memory we shouldn't have access to.

Here is a quick and dirty example using a stack-allocated `ucontext_t`, which triggers the stack-smashing detection: https://github.com/acfoltzer/ucontext-bug

On my system, `make` results in the following:

```
acfoltzer@stribog:~/src/ucontext-bug % make
cargo build
   Compiling ucontext-bug v0.1.0 (/home/acfoltzer/src/ucontext-bug)
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
cc main.c target/debug/libucontext_bug.so -o main
./main
C passing pointer to a struct of size 936
Rust writing struct of size 968
*** stack smashing detected ***: <unknown> terminated
Makefile:3: recipe for target 'run' failed
make: *** [run] Aborted (core dumped)

```

**Comment by gnzlbg** [comment]
Gotcha. AFAICT the easiest way to fix this is to revert that change, skipping ucontext in libc-test/build.rs for linux, at least until less users are using glibc < 2.28.

**Comment by acfoltzer** [comment]
Closed by #1411. Thanks!
